File: tinDB/src/main/QuerySpecification/Query/QueryGraph.java
package Query;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

// class for specifying the query in form of a graph. it consists of QueryNodes and QueryEdges.
// so far we can add nodes or edges to it and print it.
public class QueryGraph {
    public Set<QueryNode> nodes;

    public QueryGraph() {
        nodes = new HashSet<>();
    }

    public void addQueryObjectNode(QueryNode... n) {
        nodes.addAll(Arrays.asList(n));
    }

    public void addQueryObjectEdge(QueryNode source, QueryNode target, String label) {

        // .add on a hashSet does not add an element that is already present! We don't need to check whether the node is already created.

        nodes.add(source);
        nodes.add(target);

        // if this edge is already present -> don't add it.

        for (QueryEdge edge : source.edges) {
            if (edge.source == source && edge.target == target && edge.label.equalsIgnoreCase(label)) {
                return;
            }
        }
        source.edges.add(new QueryEdge(source, target, label));
    }
    public Set<QueryNode> getNodes(){
        return nodes;
    }

    // TODO: add isolated nodes! (same as in the transducer)
    public void printGraph() {
        for (QueryNode node : nodes) {
            for (QueryEdge edge : node.edges) {
                edge.print();
            }
        }
    }
}

File: src/main/kotlin/tin/model/query/QueryGraph.kt
package tin.model.query

import tin.model.transducer.TransducerNode
import kotlin.collections.HashSet

class QueryGraph {
    var nodes: MutableSet<QueryNode> = HashSet()
    var alphabet: Set<String> = HashSet()

    fun addQueryNodes(vararg n: QueryNode) {
        nodes.addAll(listOf(*n))
    }

    /**
     * Note that we cannot use "nodes.add()" or similar functions that use "equals()" here.
     * This is because we call this function to populate the graph;
     * i.e. when calling this function several times after another
     * even the same node differs now from the last iteration because one edge was added.
     * We thus have to treat the comparison of nodes differently for the sake of keeping it unique outside the graph population.
     */
    fun addQueryEdge(source: QueryNode, target: QueryNode, label: String) {
        /** check for existing source and target */
        val existingSource = findNodeWithoutEdgeComparison(source)
        val existingTarget = findNodeWithoutEdgeComparison(target)

        /** if it wasn't found: we add the new nodes.*/
        if (existingSource == null) {
            nodes.add(source)
        }

        if (existingTarget == null) {
            nodes.add(target)
        }

        val newEdge = QueryEdge(source, target, label)

        // don't add duplicate edges!
        for (existingEdge in source.edges) {
            if (existingEdge == newEdge) {
                return
            }
        }
        source.edges.add(newEdge)
    }

    /** helper function to use in the context of populating the graph only.
     * Note that since we cannot have different nodes with the same identifier,
     * we just have to check the identifier property and no other properties.
     */
    private fun findNodeWithoutEdgeComparison(queryNode: QueryNode): QueryNode? {
        return nodes.find {
            it.identifier == queryNode.identifier
        }
    }

    fun printGraph() {
        for (node in nodes) {
            for (edge in node.edges) {
                edge.print()
            }
        }
    }


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is QueryGraph) return false

        return nodes == other.nodes &&
                alphabet == other.alphabet
    }

    override fun hashCode(): Int {
        var result = nodes.hashCode()
        result = 31 * result + alphabet.hashCode()
        return result
    }


}